texar/modules/connectors/connectors.py

```diff
@@ -0,0 +1,786 @@
+    transforms samples into specified size.
+    Reparameterization allows gradients to be back-propagated through the
+    stochastic samples. Used in, e.g., Variational Autoencoders (VAEs).
+
+    Example:
+
+        .. code-block:: python
+
+            cell = LSTMCell(num_units=256)
+            # cell.state_size == LSTMStateTuple(c=256, h=256)
+            connector = ReparameterizedStochasticConnector(cell.state_size)
+            kwargs = {
+                'loc': torch.zeros([batch_size, 10]),
+                'scale_diag': torch.stack(
+                    [torch.diag(x) for x in torch.ones([batch_size, 10])],
+                    0)
+            }
+            output, sample = connector(distribution_kwargs=kwargs)
+            # output == LSTMStateTuple(c=tensor_of_shape_(batch_size, 256),
+            #                          h=tensor_of_shape_(batch_size, 256))
+            # sample == Tensor([batch_size, 10])
+            kwargs = {
+                'loc': torch.zeros([10]),
+                'scale_diag': torch.stack(
+                    [torch.diag(x) for x in torch.ones([10])], dim=0)
+            }
+            output_, sample_ = connector(distribution_kwargs=kwargs,
+                                         num_samples=batch_size_)
+            # output_ == LSTMStateTuple(c=tensor_of_shape_(batch_size_, 256),
+            #                           h=tensor_of_shape_(batch_size_, 256))
+            # sample == Tensor([batch_size_, 10])
+
+    Args:
+        output_size: Size of output **excluding** the batch dimension. For
+            example, set ``output_size`` to ``dim`` to generate output of
+            shape ``[batch_size, dim]``.
+            Can be an ``int``, a tuple of ``int``, a ``torch.Size``, or
+            a ``tuple`` of ``torch.Size``.
+            For example, to transform inputs to have decoder state size, set
+            :python:`output_size=decoder.state_size`.
+        hparams (dict, optional): Hyperparameters. Missing
+            hyperparamerter will be set to default values. See
+            :meth:`default_hparams` for the hyperparameter sturcture and
+            default values.
+        distribution: A instance of subclass of
+            :torch:`distributions.distribution.Distribution`,
+            Can be a class, its name or module path, or a class instance.
+        distribution_kwargs (dict, optional): Keyword arguments for the
+            distribution constructor. Ignored if :attr:`distribution` is a
+            class instance.
+    """
+
+    def __init__(self,
+                 output_size: OutputSize,
+                 hparams: HParamsType = None,
+                 distribution: Union[Type[T], T, str] = 'MultivariateNormal',
+                 distribution_kwargs: Optional[Dict[str, Any]] = None,):
+        ConnectorBase.__init__(self, output_size, hparams)
+        if distribution_kwargs is None:
+            distribution_kwargs = {}
+        self._dstr = utils.check_or_get_instance(
```

Comment:
Reparameterized distributions work in a different way in PyTorch. For those distributions, it is required to pass the parameters as arguments in the constructor (see `MultivariateNormal` for instance: it is required to pass `loc` (mean), and some representation of variance). Thus, the parameters must be instantiated in this `__init__` method, and stored as parameters of this module.

I checked all reparameterizable distributions in PyTorch and their constructors are pretty different. The required parameters have different names (and some has constraints on the initial values, e.g. has to be positive or positive definite), and there is no way to programmatically know the parameter names or required shapes. We may need to rethink how this module should be implemented.

One idea I have now is: Add support for automatic construction of the most common distributions (`MultivariateNormal`, and maybe a couple of others?), and for the rest, ask the user to manually add required arguments in `distribution_kwargs`. We might not need the user to construct the tensor for us, but rather have them specify the size and init_fn.

Reply:
Or, have the user manually construct the distribution. After that, we can extract the tensors within the distribution with some dirty hacks, wrap them as `nn.Parameter`s, register them as parameters of this module, and replace them back.

Reply:
I would say the second method is better, we can't cover all the reparameterizable distributions very uniformly.